### PR TITLE
New tables: Azure instance metadata

### PR DESCRIPTION
**Pull request by woodruffw**

### CODE REVIEW

File: specs/azure_instance_metadata.table [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    Column("os_type", TEXT, "Linux or Windows"),
+    Column("platform_update_domain", TEXT, "Update domain the VM is running in"),
+    Column("platform_fault_domain", TEXT, "Fault domain the VM is running in"),
+    Column("vm_id", TEXT, "Unique identifier for the VM", index=True),
+    Column("vm_size", TEXT, "VM size"),
+    Column("subscription_id", TEXT, "Azure subscription for the VM"),
+    Column("resource_group_name", TEXT, "Resource group for the VM"),
+    Column("placement_group_id", TEXT, "Placement group for the VM scale set"),
+    Column("vm_scale_set_name", TEXT, "VM scale set name"),
+    Column("zone", TEXT, "Availability zone of the VM"),
```
Comment (mossberg) [inline specs/azure_instance_metadata.table]: i noticed there are a couple pieces of data that the API can return that are unused here? (would need to update the API version date). any particular reason for not including them?

https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service#instance-metadata-data-categories

File: osquery/tables/cloud/azure/azure_instance_tags.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    TLOG << "Couldn't fetch metadata: " << s.what();
+  }
+
+  auto tags_str = tree_get(tree, "tags");
+  auto vm_id = tree_get(tree, "vmId");
+  std::vector<std::string> tags;
+
+  boost::split(tags, tags_str, boost::is_any_of(";"));
+
+  for (auto& tag : tags) {
```
Comment (mossberg) [inline osquery/tables/cloud/azure/azure_instance_tags.cpp]: nit, const auto could be here

File: osquery/tables/cloud/azure/azure_instance_tags.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    Row r;
+
+    auto colon = tag.find_first_of(':');
+
+    // This shouldn't ever happen, but it doesn't hurt to be safe.
+    if (colon == std::string::npos) {
```
Comment (mossberg) [inline osquery/tables/cloud/azure/azure_instance_tags.cpp]: nice check

File: osquery/tables/cloud/azure/azure_instance_tags.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      continue;
+    }
+
+    auto key = tag.substr(0, colon);
+    auto value = tag.substr(colon + 1);
+
+    r["vm_id"] = vm_id;
```
Comment (mossberg) [inline osquery/tables/cloud/azure/azure_instance_tags.cpp]: nit,  some std::move opportunities

File: specs/azure_instance_metadata.table (hunk 1 above)
Comment (woodruffw) [reply specs/azure_instance_metadata.table]: > any particular reason for not including them?

I didn't include them because they weren't present in all Azure regions (the table at the top of that page). OTOH, since `get` can return a default value, I don't see any particular harm in using the latest API version and providing sensible defaults for those missing keys 🙂

Comment (mossberg) [reply specs/azure_instance_metadata.table]: Ah. I thought the API version `2018-04-02` could be used since it's region is "All Generally Available Global Azure Regions" but now I see what you mean (It's not available in those three other rows).

it looks like there is some other metadata like ipv4 addresses, subnet, mac address etc that it available with the current API version. i'll plan to add those to the table also. 

for the two entries ("plan" and "publicKeys") that are only available with the latest API, I'll take a look at providing a default value.

Comment (mossberg) [reply specs/azure_instance_metadata.table]: after looking into this a bit more, i realized a few things. the outstanding attributes are currently `plan`, `pubKeys`, and the `network` json object. `pubKeys` and `network` and lists, so in order to represent them as tables, we would need to create new tables. that's a bit involved, and will delay the time to ship this feature, so i'd say let's hold off for now. the `plan` attribute is only available with the latest API version, so if we want to make sure the table works when run in any azure region, we would need to do two requests -- one with the new version, and another with a fallback version in case the first returned an invalid version error. that's also a decent bit of extra code, and it's not immediately apparent how much useful the `plan` field is anyway, so I think we should opt to hold off on this too. 

in summary, i think this PR is fine for now as far as data retrieved, and we can iterate later to add new tables for `pubKeys` / `network` or the `plan` fields if needed.

**Comment by mossberg** [comment]
https://github.com/facebook/osquery/pull/5434
